At the begining have to import the libraries that you would require. You can also include it later in the notebook.

In [2]:
import pandas as pd
import numpy as np

#Importing the dataset
We will now read the data files. 
Combats.csv contains the battle data between two pokemons.
Pokemon_data contains the properties of each pokemon. Each pokemon has some quantifiable abilities and these abilities are stored in this csv file.

the pokemon_data.csv contains indexing so we specify index_col as 0 telling python that it should not add indexing of its own and use the indexing that is present in the csv file

In [21]:
pokemon = pd.read_csv("pokemons_data.csv", index_col=0)
combats = pd.read_csv("combats.csv")

In [4]:
pokemon.head(1)

Name  Hit Points  Attack  Defense  Sp. Atk  Sp. Def  Speed  Legendary
#                                                                            
1  Bulbasaur          45      49       49       65       65     45          0

In [5]:
pokemon.shape

(800, 8)

In [6]:
combats.shape

(50000, 3)

In [7]:
combats.head(2)

First_pokemon  Second_pokemon  Winner
0            266             298     298
1            702             701     701

###Better understanding of dataset
For better understanding we will convert each number in the combats dataset to their respective pokemon names, so that it is easier for us to see.
But the machine understand numbers better as compared to strings. So this changes are only made on a temporary dataframe. We will not change the original combats dataframe.
To make a copy such that any changes would not affect the original dataframe we use copy() method

DataFrame.columns returns the names of the columns of that dataframe

In [8]:
cols = combats.columns
cols

Index(['First_pokemon', 'Second_pokemon', 'Winner'], dtype='object')

We would replace all the numbers with the name using replace method. This would change the enteries in each columns.
An example to change the columns of the new dataframe is shown below

understand_combats['First_pokemon'] = understand_combats['First_pokemon'].replace(pokemon.Name)

so we would loop through the 'cols' that contains the name of each column in the combats dataframe and change the numbers to names.

In [13]:
understand_combats = combats.copy()
for c in cols:
  understand_combats[c] = understand_combats[c].replace(pokemon.Name)

In [14]:
understand_combats.head(3)

First_pokemon Second_pokemon     Winner
0      Larvitar        Nuzleaf    Nuzleaf
1      Virizion      Terrakion  Terrakion
2       Togetic       Beheeyem   Beheeyem

The above process can be done in more pythonic way. It is shown in the below code shell. 

Rather than looping through each column value we can specify the column names that are to be changed and python would do that for us. This would remove the requirement of a loop and thus would reduce overall lines of code.

In [17]:
# understand_combats2 = combats.copy()
# understand_combats2 = understand_combats2[cols].replace(pokemon.Name)

In [18]:
# understand_combats2.head(3)

First_pokemon Second_pokemon     Winner
0      Larvitar        Nuzleaf    Nuzleaf
1      Virizion      Terrakion  Terrakion
2       Togetic       Beheeyem   Beheeyem

#Processing the dataset
###Classification probelm
This is a classification problem where we have to predict if the first pokemon won the battle or the second pokemon won the battle. So we can say that this is a binary classification problem because the outcome can be only one of the two. To represnet this binary classification we would change the winner column to 0 or 1, where 0 means that the first pokemon won and 1 means that the second pokemon won.


We loop through each row in the dataframe and check if the winner is first pokemon or not. If the winner is the first pokemon then we change the winner column to 0 or else we change it to 1 representing that the second pokemon has won the battle.

In [19]:
for i in range(len(combats)):
  if combats.Winner[i] == combats.First_pokemon[i]:
    combats.Winner[i] = 0
  else:
    combats.Winner[i] = 1

In [20]:
combats.head()

First_pokemon  Second_pokemon  Winner
0            266             298       1
1            702             701       1
2            191             668       1
3            237             683       1
4            151             231       0

The above task can be done in more pythonic way. This method is shown below

First we equate winner to the first pokemon column. this would give a list of true and false. the indexes that have the winner as first pokemon would be true and other indexes would have the false value. Now we insert this 1d matrix to combats.Winner column. This would replace the indexes containing true with the value assigned, in this case it is 0.

next we check if the winner is equal to the second pokeom. this would again give a 1d matrix of true and false. true for the cases in which the second pokemon has won. now we give this 1d matrix to combats.Winner which would change the winner column with 1 wherever there is true.

Uncomment the code shells to exexute and see the changes yourself.

In [ ]:
# combats.Winner == combats.First_pokemon

In [ ]:
# combats.Winner == combats.Second_pokemon

In [25]:
combats.Winner[combats.Winner == combats.First_pokemon] = 0
combats.Winner[combats.Winner == combats.Second_pokemon] = 1

In [26]:
combats.head(10)

First_pokemon  Second_pokemon  Winner
0            266             298       1
1            702             701       1
2            191             668       1
3            237             683       1
4            151             231       0
5            657             752       0
6            192             134       1
7             73             545       1
8            220             763       1
9            302              31       1

Our combats file is almost ready. Now we have to map it to the stats of each pokemon. For this we need to convert the pokemon dataframe into a dictonary(key:value) where key would be the number (for 266 pokemon) and value will be its stats.

We consider all the columns except the Name column, because name is not numerical and cannot be compared

In [27]:
stats = pokemon.columns
stats = stats[1:]
stats

Index(['Hit Points', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed',
       'Legendary'],
      dtype='object')

To convert the dataframe into dictonary we will use to_dict method. this method changes the column name as key and the column values as value thus forming a key:value pair.

So we need to transpose the dataframe. To transpose the dataframe use .T operator to transpose. Example is shown below.

In [28]:
pokemon[stats].T

#           1    2    3    4    5    6    ...  795  796  797  798  799  800
Hit Points   45   60   80   80   39   58  ...  108   50   50   80   80   80
Attack       49   62   82  100   52   64  ...  100  100  160  110  160  110
Defense      49   63   83  123   43   58  ...  121  150  110   60   60  120
Sp. Atk      65   80  100  122   60   80  ...   81  100  160  150  170  130
Sp. Def      65   80  100  120   50   65  ...   95  150  110  130  130   90
Speed        45   60   80   80   65   80  ...   95   50  110   70   80   70
Legendary     0    0    0    0    0    0  ...    1    1    1    1    1    1

[7 rows x 800 columns]

Now we would apply to_dict method. key would be 1,2,3...800. and the values would be the mentioned stats. we want this stats to be in a list so inside the to_dict method we mention "list". So the final output would be a key:value pair where value contains a list of stats for each pokemon

In [29]:
stats_dict = pokemon[stats].T.to_dict("list")
stats_dict

{1: [45, 49, 49, 65, 65, 45, 0],
 2: [60, 62, 63, 80, 80, 60, 0],
 3: [80, 82, 83, 100, 100, 80, 0],
 4: [80, 100, 123, 122, 120, 80, 0],
 5: [39, 52, 43, 60, 50, 65, 0],
 6: [58, 64, 58, 80, 65, 80, 0],
 7: [78, 84, 78, 109, 85, 100, 0],
 8: [78, 130, 111, 130, 85, 100, 0],
 9: [78, 104, 78, 159, 115, 100, 0],
 10: [44, 48, 65, 50, 64, 43, 0],
 11: [59, 63, 80, 65, 80, 58, 0],
 12: [79, 83, 100, 85, 105, 78, 0],
 13: [79, 103, 120, 135, 115, 78, 0],
 14: [45, 30, 35, 20, 20, 45, 0],
 15: [50, 20, 55, 25, 25, 30, 0],
 16: [60, 45, 50, 90, 80, 70, 0],
 17: [40, 35, 30, 20, 20, 50, 0],
 18: [45, 25, 50, 25, 25, 35, 0],
 19: [65, 90, 40, 45, 80, 75, 0],
 20: [65, 150, 40, 15, 80, 145, 0],
 21: [40, 45, 40, 35, 35, 56, 0],
 22: [63, 60, 55, 50, 50, 71, 0],
 23: [83, 80, 75, 70, 70, 101, 0],
 24: [83, 80, 80, 135, 80, 121, 0],
 25: [30, 56, 35, 25, 35, 72, 0],
 26: [55, 81, 60, 50, 70, 97, 0],
 27: [40, 60, 30, 31, 31, 70, 0],
 28: [65, 90, 65, 61, 61, 100, 0],
 29: [35, 60, 44, 40, 54, 55,

Now we have our dictonary ready and for the next step we have to map this dictonary to the combats dataframe. we would access each pokemon's stat using its number as the key.

map method would do this for us. It takes a dictionary as an input and applies it to the calling dataframe. to here combats.First_pokemon is the calling dataframe and we apply map(stats_dict) to this sub-dataframe.

50000 entries are changes from just the pokemon number to their stats.

In [30]:
one = combats.First_pokemon.map(stats_dict)
one

0             [50, 64, 50, 45, 50, 41, 0]
1           [91, 90, 72, 90, 129, 108, 1]
2            [55, 40, 85, 80, 105, 40, 0]
3             [40, 40, 40, 70, 40, 20, 0]
4           [70, 60, 125, 115, 70, 55, 0]
                       ...               
49995    [100, 120, 100, 150, 120, 90, 1]
49996         [60, 85, 40, 30, 45, 68, 0]
49997        [60, 50, 100, 85, 70, 65, 0]
49998        [40, 30, 50, 55, 55, 100, 0]
49999     [78, 104, 78, 159, 115, 100, 0]
Name: First_pokemon, Length: 50000, dtype: object

As done in the above code shell we would do the same thing for the second pokemon column also.

In [31]:
two = combats.Second_pokemon.map(stats_dict)
two

0          [70, 70, 40, 60, 40, 60, 0]
1        [91, 129, 90, 72, 90, 108, 1]
2         [75, 75, 75, 125, 95, 40, 0]
3         [77, 120, 90, 60, 90, 48, 0]
4         [20, 10, 230, 10, 230, 5, 0]
                     ...              
49995      [30, 40, 70, 70, 25, 60, 0]
49996      [35, 55, 40, 45, 40, 60, 0]
49997     [73, 115, 60, 60, 60, 90, 0]
49998      [25, 35, 70, 95, 55, 45, 0]
49999      [70, 80, 50, 35, 35, 35, 0]
Name: Second_pokemon, Length: 50000, dtype: object

ONE is for the first pokemon and TWO is for the second pokemon.
Now we have to compare the stats of ONE and TWO, in simple term we can say that we have to take a difference. For that we have to subtract "hit point" of senond pokemon from "hit point" of first pokemon. We have to subtract "defence" of second pokemon from "defence" of first pokemon and so on for each stats value. This is known as element-wise subtraction.

Element-wise mathematical operations can be done using numpy library. So just before applying the substraction operation we have to convert the stats for each of the pokemon into a numpy array.

Here we loop starting from 0 to 49999 and compare(subtract) TWO from ONE and save it in a list named as comparision. We define an empty list with [] and we add new data into the list using append method. So finally we have a list that contains numpy array of comparision between first and second pokemon.

In [32]:
comparision = []
for i in range(len(one)):
  comparision.append(np.array(one[i]) - np.array(two[i]))

In [33]:
comparision

[array([-20,  -6,  10, -15,  10, -19,   0]),
 array([  0, -39, -18,  18,  39,   0,   0]),
 array([-20, -35,  10, -45,  10,   0,   0]),
 array([-37, -80, -50,  10, -50, -28,   0]),
 array([  50,   50, -105,  105, -160,   50,    0]),
 array([ -10,   -3, -100,    7, -100,    5,    0]),
 array([-25,   0,  10, -45, -50, -25,   0]),
 array([-80, -20, -70, -65, -85, -55,  -1]),
 array([  0,  12,  28, -23, -28, -29,   0]),
 array([  5, -25, -10,   5, -20,  -5,   0]),
 array([ 10,  10, -25, -40, -25, -25,   0]),
 array([53, 99,  5, 17, 25, 78,  1]),
 array([-50, -90, -35, -60, -65, -30,   0]),
 array([-25, -15,  15,  15, -10,  25,   0]),
 array([ 30,  70, 100,   1,  70,   5,   0]),
 array([-27, -32,  11,   3,  12, -21,   0]),
 array([ 40,   0,   5, -30,  -5, -55,  -1]),
 array([-55, -60, -20, -20,  15,   0,   0]),
 array([ 20,  95, -62,  90, -80,  62,   0]),
 array([-21,  45,   0,   1, -25,  41,   0]),
 array([ -1, -59, -32,  21,   6, -20,   0]),
 array([ 26,  -8, -15, -35, -30,   0,   0]),
 ar

We have the comparision. Now we would add the winner column to the comparision list. To add this Winner column we are coverting the comparision list into a pandas dataframe. Each comparision contains 6 entries(hit points, attack, defence, sp. atk, sp. def, speed and legnedry). So while creating the dataframe we will mention what are the column names for our understanding.

In [34]:
almost_data = pd.DataFrame(comparision, columns=stats)
almost_data.head()

Hit Points  Attack  Defense  Sp. Atk  Sp. Def  Speed  Legendary
0         -20      -6       10      -15       10    -19          0
1           0     -39      -18       18       39      0          0
2         -20     -35       10      -45       10      0          0
3         -37     -80      -50       10      -50    -28          0
4          50      50     -105      105     -160     50          0

Now we would use concat method to include the winner column in the dataframe. we want this data to be added as a column, so mention that the axis for this concatenation is 1(1 is for column and 0 is for rows).

Finally our dataframe is ready for splitting purpose.

In [35]:
data = pd.concat([almost_data, combats.Winner], axis=1)
data

Hit Points  Attack  Defense  Sp. Atk  Sp. Def  Speed  Legendary  Winner
0             -20      -6       10      -15       10    -19          0       1
1               0     -39      -18       18       39      0          0       1
2             -20     -35       10      -45       10      0          0       1
3             -37     -80      -50       10      -50    -28          0       1
4              50      50     -105      105     -160     50          0       0
...           ...     ...      ...      ...      ...    ...        ...     ...
49995          70      80       30       80       95     30          1       0
49996          25      30        0      -15        5      8          0       0
49997         -13     -65       40       25       10    -25          0       1
49998          15      -5      -20      -40        0     55          0       0
49999           8      24       28      124       80     65          0       0

[50000 rows x 8 columns]

Our dataframe is finally ready and has the comparision between the first pokemon and the second pokemon. Here -ve value means that the second pokemon was more powerful in the mention column and its magnitude defines how stronger he is. +ve value means that the first pokemon was stronger and the magnitude defines how stronger he is as compared to the secnod pokemon. Our last column shows which pokemon has won. 0 means that the first pokemon won and 1 means that the second pokemon won.

We would save our dataframe so that we can coninue our work from here. to_csv method is used to save the dataframe into a file. After you execute the code shell below, refresh the files column and ready.csv file would appear in the files column.

In [60]:
data.to_csv('ready.csv')

Now we can either import ready.csv file here or we can directly coninue working

In [78]:
data = pd.read_csv("ready.csv", index_col=0)
data.head()

Hit Points  Attack  Defense  Sp. Atk  Sp. Def  Speed  Legendary  Winner
0         -20      -6       10      -15       10    -19          0       1
1           0     -39      -18       18       39      0          0       1
2         -20     -35       10      -45       10      0          0       1
3         -37     -80      -50       10      -50    -28          0       1
4          50      50     -105      105     -160     50          0       0

###Feature Scaling
There are two types of scaling. Standard scaling and normalization.

Normalization is a good technique to use when you do not know the distribution of your data or when you know the distribution is not Gaussian (a bell curve). Normalization is useful when your data has varying scales and the algorithm you are using does not make assumptions about the distribution of your data, such as k-nearest neighbors and artificial neural networks.

Standardization assumes that your data has a Gaussian (bell curve) distribution. This does not strictly have to be true, but the technique is more effective if your attribute distribution is Gaussian. Standardization is useful when your data has varying scales and the algorithm you are using does make assumptions about your data having a Gaussian distribution, such as linear regression, logistic regression, and linear discriminant analysis.

[More about standardization and normalization](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Normalization%20is%20useful%20when%20your,Gaussian%20(bell%20curve)%20distribution.)

In [81]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler() 
data_scaled = scaler.fit_transform(X)
data_scaled

array([[0.45555556, 0.46857143, 0.51724138, ..., 0.52757794, 0.45074627,
        0.5       ],
       [0.5       , 0.37428571, 0.45287356, ..., 0.5971223 , 0.50746269,
        0.5       ],
       [0.45555556, 0.38571429, 0.51724138, ..., 0.52757794, 0.50746269,
        0.5       ],
       ...,
       [0.47111111, 0.3       , 0.5862069 , ..., 0.52757794, 0.43283582,
        0.5       ],
       [0.53333333, 0.47142857, 0.44827586, ..., 0.50359712, 0.67164179,
        0.5       ],
       [0.51777778, 0.55428571, 0.55862069, ..., 0.69544365, 0.70149254,
        0.5       ]])

In [82]:
data_scaled

array([[0.45555556, 0.46857143, 0.51724138, ..., 0.52757794, 0.45074627,
        0.5       ],
       [0.5       , 0.37428571, 0.45287356, ..., 0.5971223 , 0.50746269,
        0.5       ],
       [0.45555556, 0.38571429, 0.51724138, ..., 0.52757794, 0.50746269,
        0.5       ],
       ...,
       [0.47111111, 0.3       , 0.5862069 , ..., 0.52757794, 0.43283582,
        0.5       ],
       [0.53333333, 0.47142857, 0.44827586, ..., 0.50359712, 0.67164179,
        0.5       ],
       [0.51777778, 0.55428571, 0.55862069, ..., 0.69544365, 0.70149254,
        0.5       ]])

In [68]:
# stats = data.columns
# stats = stats[:-1]
# stats

Index(['Hit Points', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed',
       'Legendary'],
      dtype='object')

In [69]:
data_scaled = pd.DataFrame(data_scaled, columns=stats)
data_scaled

Hit Points    Attack   Defense   Sp. Atk   Sp. Def     Speed  Legendary
0        0.455556  0.468571  0.517241  0.450425  0.527578  0.450746        0.5
1        0.500000  0.374286  0.452874  0.543909  0.597122  0.507463        0.5
2        0.455556  0.385714  0.517241  0.365439  0.527578  0.507463        0.5
3        0.417778  0.257143  0.379310  0.521246  0.383693  0.423881        0.5
4        0.611111  0.628571  0.252874  0.790368  0.119904  0.656716        0.5
...           ...       ...       ...       ...       ...       ...        ...
49995    0.655556  0.714286  0.563218  0.719547  0.731415  0.597015        1.0
49996    0.555556  0.571429  0.494253  0.450425  0.515588  0.531343        0.5
49997    0.471111  0.300000  0.586207  0.563739  0.527578  0.432836        0.5
49998    0.533333  0.471429  0.448276  0.379603  0.503597  0.671642        0.5
49999    0.517778  0.554286  0.558621  0.844193  0.695444  0.701493        0.5

[50000 rows x 7 columns]

In [ ]:
# data_scaled.head()

In [ ]:
# from sklearn.preprocessing import StandardScaler
# sc = StandardScaler()
# X_train = sc.fit_transform(X_train)

Now we have to divide our dataset into dependent variable and independent variable. We will use 'y' for dependent variable and 'X' for the independent variables.

 We want all the rows and all the columns except the last one in X. Similarly we want all the rows and only the last column in y. Dependent variable is the one we want to predict.

In [80]:
X = data.iloc[:, :-1].values
y = data.iloc[:, -1].values

#Splitting the dataset
We will split the dataset into training set and validation set. Since we are familiar with the name 'test set' we would call the 'validation set' as 'test set' for this course. Train test and validation set is used more in Deep Learning. Machine learning projects generally deals with training and test set only.
If you want to learn more about the train test and validation test. [Follow this link](https://machinelearningmastery.com/difference-test-validation-datasets/)

In [83]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data_scaled, y, test_size = 0.25)

In [84]:
X_train

array([[0.67777778, 0.53428571, 0.46436782, ..., 0.47242206, 0.49552239,
        0.5       ],
       [0.61111111, 0.63428571, 0.83908046, ..., 0.5323741 , 0.45671642,
        0.5       ],
       [0.48888889, 0.42571429, 0.57471264, ..., 0.55155875, 0.42686567,
        0.5       ],
       ...,
       [0.54444444, 0.61428571, 0.57471264, ..., 0.43165468, 0.47761194,
        0.5       ],
       [0.48222222, 0.47714286, 0.40229885, ..., 0.35971223, 0.58208955,
        0.5       ],
       [0.48222222, 0.46857143, 0.49655172, ..., 0.5059952 , 0.59701493,
        0.5       ]])

In [85]:
X_test

array([[0.45555556, 0.51714286, 0.50574713, ..., 0.34772182, 0.45970149,
        0.5       ],
       [0.43111111, 0.15714286, 0.56321839, ..., 0.57553957, 0.40298507,
        0.5       ],
       [0.47777778, 0.45714286, 0.40229885, ..., 0.44844125, 0.45373134,
        0.5       ],
       ...,
       [0.54444444, 0.48571429, 0.43678161, ..., 0.55155875, 0.68656716,
        1.        ],
       [0.26666667, 0.2       , 0.33793103, ..., 0.42446043, 0.4358209 ,
        0.5       ],
       [0.53333333, 0.44571429, 0.54022989, ..., 0.56354916, 0.57910448,
        0.5       ]])

In [87]:
y_train

array([1, 1, 1, ..., 1, 0, 0])

In [88]:
y_test

array([1, 1, 1, ..., 0, 1, 0])

#Training
We would now apply different classification algorithms to see which will perform better.

Before this lets import evaluation techniques for comapision

In [44]:
from sklearn.metrics import accuracy_score, confusion_matrix

###Logistic Regression

In [89]:
from sklearn.linear_model import LogisticRegression
log_model = LogisticRegression()
log_model.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [90]:
log_pred = log_model.predict(X_test)
accuracy_score(y_test, log_pred)

0.89

###Decision Tree

In [91]:
from sklearn.tree import DecisionTreeClassifier
decision_model = DecisionTreeClassifier(criterion = 'entropy')
decision_model.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [92]:
decision_pred = decision_model.predict(X_test)
accuracy_score(y_test, decision_pred)

0.9132

In [93]:
confusion_matrix(y_test, decision_pred)

array([[5370,  541],
       [ 544, 6045]])

###Random Forest

In [94]:
from sklearn.ensemble import RandomForestClassifier
random_model = RandomForestClassifier(n_estimators=100)
random_model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [95]:
random_pred = random_model.predict(X_test)
accuracy_score(y_test, random_pred)

0.94728

#Performing predictions
Perform predictions for 'test.csv'